elf/module.go

```diff
@@ -504,20 +503,29 @@ func (b *Module) closeSocketFilters() error {
 	return nil
 }
 
-func unpinMap(m *Map) error {
-	if m.m.def.pinning == 0 {
-		return nil
+func unpinMap(m *Map, pinPath string) error {
+	mapPath, err := getMapPath(&m.m.def, m.Name, pinPath)
+	if err != nil {
+		return err
 	}
-	namespace := C.GoString(&m.m.def.namespace[0])
-	mapPath := filepath.Join(BPFFSPath, namespace, BPFDirGlobals, m.Name)
 	return syscall.Unlink(mapPath)
 }
 
 func (b *Module) closeMaps(options map[string]CloseOptions) error {
 	for _, m := range b.maps {
 		doUnpin := options[fmt.Sprintf("maps/%s", m.Name)].Unpin
-		if m.m.def.pinning > 0 && doUnpin {
-			unpinMap(m)
+		if doUnpin {
+			mapDef := m.m.def
+			if mapDef.pinning == PIN_CUSTOM_NS {
+				pinPath := options[fmt.Sprintf("maps/%s", m.Name)].PinPath
+				if err := unpinMap(m, pinPath); err != nil {
+					return fmt.Errorf("error unpinning path %q: %v", pinPath, err)
+				}
+			} else if mapDef.pinning == PIN_GLOBAL_NS {
+				// ignore error for global objects as pinPath would always be an empty string.
+				unpinMap(m, "")
```

Comment:
which error do you ignore? `unpinMap()` does not seem to error out on global pinning when PinPath is empty.

Reply:
@alban Yes, that's what I mean. I'll change the comment to `we don't need to check error for global objects as pinPath would always be an empty string.`

Reply:
it could still fail for another reason: for example, an user manually unpinned the map by running `/bin/rm` from a terminal. So we can still check the error in the same way as the previous case.

Reply:
@alban I see your point. To check for the precise error, we would need to distinguish errors of getMapPath from unlink errors. Maybe we would need to define a new error from getMapPath to check for that later, etc. That would complicate the source tree a little more, and would not be exactly the same direction that we took in this PR, right? So I just wrote a precise comment like below:

```
// For the moment we don't need to check error for global objects,
// as pinPath would always be an empty string. Though in the future,
// it would make sense to check for precise error codes to distinguish
// errors of getMapPath() from errors from unlink(2).
```

Reply:
@alban Ok, I just squashed this part again into a simple error checking with the help of @schu. Hopefully it would make sense.